File: solutions/Medium/Binary_Tree_Level_Order_Traversal_0102.py
from typing import List, Optional
from collections import deque

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        """
        Function: levelOrder
        Description: Performs level order traversal of a binary tree.

        Parameters:
        - root (Optional[TreeNode]): The root of the binary tree.

        Returns:
        - List[List[int]]: A list of lists, where each list contains node values at a specific level.
        """
        if not root:
            return []

        result = []
        queue = deque([root])

        while queue:
            level_size = len(queue)
            level_nodes = []
            for _ in range(level_size):
                node = queue.popleft()
                level_nodes.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            result.append(level_nodes)

        return result


# Helper function to construct a binary tree from a level-order list
def list_to_binary_tree(lst):
    """Converts a level-order list to a binary tree."""
    if not lst or lst[0] is None:
        return None

    nodes = [TreeNode(val) if val is not None else None for val in lst]

    for i, node in enumerate(nodes):
        if node:
            left_idx = 2 * i + 1
            right_idx = 2 * i + 2
            if left_idx < len(lst):
                node.left = nodes[left_idx]
            if right_idx < len(lst):
                node.right = nodes[right_idx]

    return nodes[0]  # Root node


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: Standard binary tree
    root = list_to_binary_tree([3, 9, 20, None, None, 15, 7])
    print(solution.levelOrder(root))
    # Expected output: [[3], [9, 20], [15, 7]]

    # Test case 2: Single-node tree
    root = list_to_binary_tree([1])
    print(solution.levelOrder(root))
    # Expected output: [[1]]

    # Test case 3: Empty tree
    root = list_to_binary_tree([])
    print(solution.levelOrder(root))
    # Expected output: []

    # Test case 4: Left-skewed tree
    root = list_to_binary_tree([1, 2, None, 3, None, 4, None])
    print(solution.levelOrder(root))
    # Expected output: [[1], [2], [3], [4]]

    # Test case 5: Right-skewed tree
    root = list_to_binary_tree([1, None, 2, None, 3, None, 4])
    print(solution.levelOrder(root))
    # Expected output: [[1], [2], [3], [4]]

    # Test case 6: Tree with multiple levels
    root = list_to_binary_tree(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15])
    print(solution.levelOrder(root))
    # Expected output: [[1], [2, 3], [4, 5, 6, 7], [8, 9, 10, 11, 12, 13, 14, 15]]

    print("All test cases passed!")

# Complexity Analysis
# Time Complexity: O(N), where N is the number of nodes in the binary tree. Each node is processed once.
# Space Complexity: O(N), where N is the number of nodes in the binary tree. The maximum number of nodes in a single level is N/2, which occurs at the leaf level.
# Therefore, the space complexity is O(N).

File: solutions/Medium/Binary_Tree_Level_Order_Traversal_0102.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val);
 *     this.left = (left===undefined ? null : left);
 *     this.right = (right===undefined ? null : right);
 * }
 */

/**
 * Function: levelOrder
 * Description: Given the root of a binary tree, this function returns a 
 * level order traversal of its nodes' values (i.e., left to right, level by level).
 *
 * @param {TreeNode} root - The root node of the binary tree.
 * @returns {number[][]} - A list of lists where each inner list contains values at a specific level.
 */
const levelOrder = function(root) {
    if (!root) return []; // Edge case: Empty tree

    let result = [];
    let queue = [root];

    while (queue.length > 0) {
        let levelSize = queue.length; // Number of nodes at the current level
        let levelNodes = [];

        for (let i = 0; i < levelSize; i++) {
            let currentNode = queue.shift(); // Dequeue the front node
            levelNodes.push(currentNode.val); // Add node value to current level list

            // Enqueue child nodes for the next level
            if (currentNode.left) queue.push(currentNode.left);
            if (currentNode.right) queue.push(currentNode.right);
        }

        result.push(levelNodes);
    }

    return result;
};

// Example Test Cases
if (require.main === module) {
    const TreeNode = function(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    };

    console.log("Test 1: Basic case with root = [3,9,20,null,null,15,7]");
    let root1 = new TreeNode(3, 
        new TreeNode(9), 
        new TreeNode(20, new TreeNode(15), new TreeNode(7))
    );
    console.log(levelOrder(root1));
    // Expected output: [[3],[9,20],[15,7]]

    console.log("Test 2: Single node tree");
    let root2 = new TreeNode(1);
    console.log(levelOrder(root2));
    // Expected output: [[1]]

    console.log("Test 3: Empty tree");
    console.log(levelOrder(null));
    // Expected output: []

    console.log("Test 4: Left-skewed tree");
    let root3 = new TreeNode(1, 
        new TreeNode(2, 
            new TreeNode(3, 
                new TreeNode(4, 
                    new TreeNode(5)
                )
            )
        )
    );
    console.log(levelOrder(root3));
    // Expected output: [[1],[2],[3],[4],[5]]

    console.log("Test 5: Right-skewed tree");
    let root4 = new TreeNode(1, null, 
        new TreeNode(2, null, 
            new TreeNode(3, null, 
                new TreeNode(4, null, 
                    new TreeNode(5)
                )
            )
        )
    );
    console.log(levelOrder(root4));
    // Expected output: [[1],[2],[3],[4],[5]]

    console.log("Test 6: Large tree with mixed structure");
    let root5 = new TreeNode(10, 
        new TreeNode(5, 
            new TreeNode(2), 
            new TreeNode(7)
        ), 
        new TreeNode(15, 
            new TreeNode(12), 
            new TreeNode(20)
        )
    );
    console.log(levelOrder(root5));
    // Expected output: [[10],[5,15],[2,7,12,20]]
}

/*
Time Complexity: O(n) - Each node is visited once.
Space Complexity: O(n) - Stores nodes in queue and result array.
*/

export default levelOrder;
